README
This application when used after band calculations in vasp, can be used to generate band structure files.

This application is released under the GPL v2 License


CHANGELOG
Version .25
~~~~~~~~~~~

-FEATURE:Program will now delete files under Windows (Linux support was provided previously)
-MISC:Code clean up. Variables are now defined in a less ambiguous way.
-MISC:Several functions have now been declared static
-MISC:Changelog now added to document changes between versions


src/arg.c
/*
 * arg.c
 * 
 * Copyright 2015-2016 Emma Davenport <Davenport.physics@gmail.com>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */


#include "arg.h"

#define NUM_ARGS 8

bool Debug = FALSE;
bool Tabs  = FALSE;

char SpinUp[MIN_BUFFER_LENGTH]   = "SpinUp";
char SpinDown[MIN_BUFFER_LENGTH] = "SpinDown";

double FermiEnergy = 0.0;

static void arg_all(int index, int argc, char *argv[]);
static void arg_tabs(int index, int argc, char *argv[]);
static void arg_debug(int index, int argc, char *argv[]);
static void arg_output_file(int index, int argc, char *argv[]);
static void arg_fermi_energy(int index, int argc, char *argv[]);
static void arg_auto_fermi_energy(int index, int argc, char *argv[]);


static const ArgFunctions ArgCheck[NUM_ARGS] = 
{{"-all"         , &arg_all},
 {"-debug"       , &arg_debug},
 {"-tabs"        , &arg_tabs},
 {"-output-file" , &arg_output_file},
 {"-o"           , &arg_output_file},
 {"-fermi-energy", &arg_fermi_energy},
 {"-fe"          , &arg_fermi_energy},
 {"-auto"        , &arg_auto_fermi_energy}};

void ParseArgs(int argc, char *argv[]) {

	int x, y;
	for (x = 1;x < argc;x++) {
	
		for (y = 0;y < NUM_ARGS;y++) {
		
			if (strcasecmp(argv[x], ArgCheck[y].string) == 0) {
			
				ArgCheck[y].function(x, argc, argv);
				
			}
			
		}
		
	}
	
}

void arg_all(int index, int argc, char *argv[]) {

	Tabs  = TRUE;
	Debug = TRUE;
	
	arg_auto_fermi_energy(index, argc, argv);
	
}

void arg_tabs(int index, int argc, char *argv[]) {
	
	Tabs = TRUE;
	
}

void arg_debug(int index, int argc, char *argv[]) {

	Debug = TRUE;
	
}

void arg_output_file(int index, int argc, char *argv[]) {

	if ((index + 1) == argc) {
	
		printf("Did not pass output filename\n");
		exit(EXIT_FAILURE);
		
	} else {
	
		strncpy(SpinUp, argv[index + 1], MIN_BUFFER_LENGTH);
		strncat(SpinUp, "-Up", 3);
		
		strncpy(SpinDown, argv[index + 1], MIN_BUFFER_LENGTH);
		strncat(SpinUp, "-Down", 5);
		
	}
	
}

void arg_fermi_energy(int index, int argc, char *argv[]) {

	if ( (index + 1) == argc ) {
	
		printf("Did not pass fermi energy\n");
		exit(EXIT_FAILURE);
		
	} else {
		
		FermiEnergy = atof(argv[index + 1]);
		
		if (FermiEnergy == 0.0) {
		
			printf("Did not pass a good value for the Fermi energy");
			
		} else {
			
			if (Debug == TRUE) {
			
				printf("Fermi energy given = %f\n", FermiEnergy);
				
			}
			
		}
		
	}
	
}

void arg_auto_fermi_energy(int index, int argc, char *argv[]) {

	char buffer[512];

	FILE *fp = fopen("OUTCAR", "r");

	if (fp == NULL) {

		printf("No file OUTCAR found. Halting Execution \n");
		exit(0);

	}

	while (fgets(buffer, 512, fp) != NULL) {

		if (strncmp(buffer, " E-fermi :", 10) == 0) {

			char temp[6];

			temp[0] = buffer[13];
			temp[1] = buffer[14];
			temp[2] = buffer[15];
			temp[3] = buffer[16];
			temp[4] = buffer[17];
			temp[5] = buffer[18];

			FermiEnergy = atof(temp);

			if (Debug == TRUE) {

				printf("Extracted Fermi energy = %g\n", FermiEnergy);

			}

			if (FermiEnergy == 0.0) {

				printf("Failed to extract Fermi energy from OUTCAR\n");

			}

			break;
		}

	}

	fclose(fp);
}


src/arg.h
/*
 * arg.h
 * 
 * Copyright 2015-2016 Emma Davenport <Davenport.physics@gmail.com>
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * 
 * 
 */

#ifndef ARG_H
#define ARG_H

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MIN_BUFFER_LENGTH 256

typedef enum {

	FALSE = 0,
	TRUE

} bool;

typedef struct {
	
	char *string;
	void (*function)(int index,int argc, char *argv[]);
	
} ArgFunctions;


extern bool Debug;
extern bool Tabs;

extern char SpinUp[MIN_BUFFER_LENGTH];
extern char SpinDown[MIN_BUFFER_LENGTH];

extern double FermiEnergy;

/*
 * Arg parsing function
 * 
 * */
void ParseArgs(int argc, char *argv[]); 

#endif



src/eigenval.c
/*
 * eigenval.c
 *
 * Copyright 2015-2016 Emma Davenport <Davenport.physics@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 *
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <malloc.h>
#include <string.h>
#include <unistd.h>

#include "arg.h"

static char *SOFTWARE_VERSION = ".25";
static char *Filename         = "EIGENVAL";

static float *SpinUpColumn;
static float *SpinDownColumn;

static int NumberOfBands = 0;
static int NumberOfGrids = 0;

static FILE *fp;

static void initialize(int argc, char *argv[]);
static void check_for_existing_files();
static void delete_existing_file(char File[MIN_BUFFER_LENGTH]);
static void ReadData(int iteration);
static void Transpose(int iteration);

int main(int argc, char *argv[]) {

	printf("C-BS-VASP Version %s\n", SOFTWARE_VERSION);
	printf("Make sure that you are performing spin polarization calculations. i.e ISPIN=2\n");

	initialize(argc, argv);

	int x;
	for (x = 0;x < NumberOfGrids;x++) {

		ReadData((x + 1));
		Transpose((x + 1));

	}

	free(SpinUpColumn);
	free(SpinDownColumn);

	fclose(fp);

	return 0;
}

static char *RegularFloatString = "%.4f";
static char *TabsFloatString    = "%.4f\t";
static char *CommaFloatString   = "%.4f,";
static char *FloatString;

static char *TabsIntString  = "%d\t";
static char *CommaIntString = "%d,";
static char *IntString;

void initialize(int argc, char *argv[]) {

	ParseArgs(argc, argv);
	
	fp = fopen(Filename, "r");

	if (fp == NULL) {

		printf("%s file not found\n", Filename);
		exit(EXIT_FAILURE);

	}
	
	check_for_existing_files();

	char Buffer[MIN_BUFFER_LENGTH];

	int x;
	for (x = 0;x < 5;x++)
		fgets(Buffer, MIN_BUFFER_LENGTH, fp);
	fscanf(fp, "%*d%d%d\n", &NumberOfGrids, &NumberOfBands);

	if (Debug == TRUE)
		printf("Number of grids (Related to KPOINTS) = %d, Number of Bands = %d\n", NumberOfGrids, NumberOfBands);

	SpinUpColumn  = malloc(NumberOfBands * sizeof(float));
	SpinDownColumn = malloc(NumberOfBands * sizeof(float));
	
	if (Tabs == TRUE) {
	
		FloatString = TabsFloatString;
		IntString   = TabsIntString;
		
	} else {
		
		FloatString = CommaFloatString;
		IntString   = CommaIntString;
		
	}

}
void check_for_existing_files() {

	FILE *localfp = fopen(SpinUp, "r");
	
	if (localfp != NULL) {
	
		fclose(localfp);
		delete_existing_file(SpinUp);
		
	}
	
	localfp = fopen(SpinDown, "r");
	if (localfp != NULL) {
		
		fclose(localfp);
		delete_existing_file(SpinDown);
		
	}
	
}

void delete_existing_file(char File[MIN_BUFFER_LENGTH]) {
	
	int code = unlink(File);
	
	if (code == 0) {
	
		printf("Deleted old %s successfully.\n", File);
		
	} else if (code == 1) {
	
		printf("There was an error deleting file %s.\n", File);
		printf("Please delete the file manually.\n");
		
	}
	
}

void ReadData(int iteration) {

	char Buffer[MIN_BUFFER_LENGTH];

	/*
	 * This conditional statement fixes a bug that I do not want
	 * to spend time diagnosing properly.
	 *
	 * */
	if (iteration == 1) {

		fgets(Buffer, MIN_BUFFER_LENGTH, fp);

	} else {

		fgets(Buffer, MIN_BUFFER_LENGTH, fp);
		fgets(Buffer, MIN_BUFFER_LENGTH, fp);
		fgets(Buffer, MIN_BUFFER_LENGTH, fp);

	}

	int x;
	for (x = 0;x < NumberOfBands;x++) {

		fscanf(fp, "%*d%f%f", &SpinUpColumn[x], &SpinDownColumn[x]);

		if (FermiEnergy != 0.0) {

			SpinUpColumn[x]  -= FermiEnergy;
			SpinDownColumn[x] -= FermiEnergy;

		}

	}

}

void Transpose(int iteration) {

	FILE *SpinUpBand   = fopen(SpinUp, "a");
	FILE *SpinDownBand = fopen(SpinDown, "a");

	fprintf(SpinUpBand, IntString, iteration);
	fprintf(SpinDownBand, IntString, iteration);

	int x;
	for (x = 0;x < NumberOfBands;x++) {

		if ((x + 1) == NumberOfBands) {
				
			fprintf(SpinUpBand, RegularFloatString, SpinUpColumn[x]);
			fprintf(SpinDownBand, RegularFloatString, SpinDownColumn[x]);
				
		} else {
				
			fprintf(SpinUpBand, FloatString, SpinUpColumn[x]);
			fprintf(SpinDownBand, FloatString, SpinDownColumn[x]);
				
		}

	}
	fprintf(SpinUpBand, "\n");
	fprintf(SpinDownBand, "\n");


	fclose(SpinUpBand);
	fclose(SpinDownBand);

}
